contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -0,0 +1,443 @@
+
+import "@0x/contracts-erc20/contracts/src/v06/IEtherTokenV06.sol";
+import "@0x/contracts-utils/contracts/src/v06/LibSafeMathV06.sol";
+import "@0x/contracts-utils/contracts/src/v06/LibMathV06.sol";
+import "../errors/LibNativeOrdersRichErrors.sol";
+import "../fixins/FixinCommon.sol";
+import "../fixins/FixinEIP712.sol";
+import "../fixins/FixinTokenSpender.sol";
+import "../migrations/LibMigrate.sol";
+import "../storage/LibNativeOrdersStorage.sol";
+import "../storage/LibOtcOrdersStorage.sol";
+import "./interfaces/IFeature.sol";
+import "./interfaces/IOtcOrdersFeature.sol";
+import "./libs/LibNativeOrder.sol";
+import "./libs/LibSignature.sol";
+
+
+/// @dev Feature for interacting with OTC orders.
+contract OtcOrdersFeature is
+    IFeature,
+    IOtcOrdersFeature,
+    FixinCommon,
+    FixinEIP712,
+    FixinTokenSpender
+{
+    using LibSafeMathV06 for uint256;
+
+    /// @dev Name of this feature.
+    string public constant override FEATURE_NAME = "OtcOrders";
+    /// @dev Version of this feature.
+    uint256 public immutable override FEATURE_VERSION = _encodeVersion(1, 0, 0);
+    /// @dev The WETH token contract.
+    IEtherTokenV06 private immutable WETH;
+
+    constructor(address zeroExAddress, IEtherTokenV06 weth)
+        public
+        FixinEIP712(zeroExAddress)
+    {
+        WETH = weth;
+    }
+
+    /// @dev Initialize and register this feature.
+    ///      Should be delegatecalled by `Migrate.migrate()`.
+    /// @return success `LibMigrate.SUCCESS` on success.
+    function migrate()
+        external
+        returns (bytes4 success)
+    {
+        _registerFeatureFunction(this.fillOtcOrder.selector);
+        _registerFeatureFunction(this.fillTakerSignedOtcOrder.selector);
+        _registerFeatureFunction(this.getOtcOrderInfo.selector);
+        _registerFeatureFunction(this.getOtcOrderHash.selector);
+        _registerFeatureFunction(this.minTxOriginNonce.selector);
+        return LibMigrate.MIGRATE_SUCCESS;
+    }
+
+    /// @dev Fill an OTC order for up to `takerTokenFillAmount` taker tokens.
+    /// @param order The OTC order.
+    /// @param makerSignature The order signature from the maker.
+    /// @param takerTokenFillAmount Maximum taker token amount to fill this
+    ///        order with.
+    /// @param unwrapWeth Whether or not to unwrap bought WETH into ETH
+    ///        before transferring it to the taker. Should be set to false
+    ///        if the maker token is not WETH.
+    /// @return takerTokenFilledAmount How much taker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillOtcOrder(
+        LibNativeOrder.OtcOrder memory order,
+        LibSignature.Signature memory makerSignature,
+        uint128 takerTokenFillAmount,
+        bool unwrapWeth
+    )
+        public
+        override
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        if (!_isSenderValidTaker(order.taker)) {
```

Comment:
Isn't this check done twice? once here, and then another time in the `_fillOtcOrderPrivate`? Is there a clean way to avoid running the function twice?

Reply:
(I'm mentioning this since I assume it consumes gas, however I'm unsure what gas savings a refactor will have)

Reply:
EIP-2929 made warm SLOADs cheap